File: gnd/src/main/java/com/google/android/gnd/model/locationofinterest/LocationOfInterest.java
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.model.locationofinterest;

import androidx.annotation.NonNull;
import com.google.android.gnd.model.AuditInfo;
import com.google.android.gnd.model.Survey;
import com.google.android.gnd.model.job.Job;
import com.google.android.gnd.model.mutation.LocationOfInterestMutation;
import com.google.android.gnd.model.mutation.Mutation.SyncStatus;
import com.google.android.gnd.model.mutation.Mutation.Type;
import java.util.Date;
import javax.annotation.Nullable;
import javax.annotation.OverridingMethodsMustInvokeSuper;

/** Base class for user-defined locations of interest shown on map. */
public abstract class LocationOfInterest<B extends LocationOfInterest.Builder> {
  public boolean isPoint() {
    return this instanceof PointOfInterest;
  }

  public boolean isGeoJson() {
    return this instanceof GeoJsonLocationOfInterest;
  }

  public boolean isPolygon() {
    return this instanceof PolygonOfInterest;
  }

  @NonNull
  public abstract String getId();

  public abstract Survey getSurvey();

  public abstract Job getJob();

  @Nullable
  public abstract String getCustomId();

  @Nullable
  public abstract String getCaption();

  /**
   * Returns the user and time audit info pertaining to the creation of this location of interest.
   */
  public abstract AuditInfo getCreated();

  /**
   * Returns the user and time audit info pertaining to the last modification of this location of
   * interest.
   */
  public abstract AuditInfo getLastModified();

  @OverridingMethodsMustInvokeSuper
  public LocationOfInterestMutation toMutation(Type type, String userId) {
    return LocationOfInterestMutation.builder()
        .setType(type)
        .setSyncStatus(SyncStatus.PENDING)
        .setSurveyId(getSurvey().getId())
        .setLocationOfInterestId(getId())
        .setJobId(getJob().getId())
        .setUserId(userId)
        .setClientTimestamp(new Date())
        .build();
  }

  public abstract static class Builder<T extends LocationOfInterest.Builder> {
    // TODO: Use newFoo or foo consistently.
    public abstract T setId(@NonNull String newId);

    public abstract T setSurvey(@NonNull Survey survey);

    public abstract T setJob(@NonNull Job newJob);

    public abstract T setCustomId(@Nullable String newCustomId);

    public abstract T setCaption(@Nullable String newCaption);

    public abstract T setCreated(@NonNull AuditInfo newCreated);

    public abstract T setLastModified(@NonNull AuditInfo newLastModified);
  }
}

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.ground.model.locationofinterest

import com.google.android.ground.model.AuditInfo
import com.google.android.ground.model.geometry.Geometry
import com.google.android.ground.model.job.Job
import com.google.android.ground.model.mutation.LocationOfInterestMutation
import com.google.android.ground.model.mutation.Mutation
import com.google.android.ground.model.mutation.Mutation.SyncStatus

/** Alias for a map of properties with string names. */
typealias LoiProperties = Map<String, Any>

const val LOI_NAME_PROPERTY = "name"

fun generateProperties(loiName: String? = null): LoiProperties =
  loiName?.let { mapOf(LOI_NAME_PROPERTY to it) } ?: mapOf()

/** User-defined locations of interest (LOI) shown on the map. */
data class LocationOfInterest(
  /** A system-defined ID for this LOI. */
  val id: String,
  /** The survey ID associated with this LOI. */
  val surveyId: String,
  /** The job associated with this LOI. */
  val job: Job,
  /** A user-specified ID for this location of interest. */
  val customId: String = "",
  /** User and time audit info pertaining to the creation of this LOI. */
  val created: AuditInfo,
  /** User and time audit info pertaining to the last modification of this LOI. */
  val lastModified: AuditInfo,
  /** Geometry associated with this LOI. */
  val geometry: Geometry,
  /** The number of submissions that have been made for this LOI. */
  val submissionCount: Int = 0,
  /** Custom map of properties for this LOI. Corresponds to the properties field in GeoJSON */
  val properties: LoiProperties = mapOf(),
  /** Whether the LOI was predefined in the survey or not (i.e. added ad hoc). */
  val isPredefined: Boolean? = null,
) {

  /**
   * Converts this LOI to a mutation that can be used to update this LOI in the remote and local
   * database.
   */
  // TODO: Remove this test-only method
  // Issue URL: https://github.com/google/ground-android/issues/2903
  fun toMutation(type: Mutation.Type, userId: String): LocationOfInterestMutation =
    LocationOfInterestMutation(
      jobId = job.id,
      type = type,
      syncStatus = SyncStatus.PENDING,
      surveyId = surveyId,
      locationOfInterestId = id,
      userId = userId,
      customId = customId,
      clientTimestamp = lastModified.clientTimestamp,
      geometry = geometry,
      submissionCount = submissionCount,
      properties = properties,
      isPredefined = isPredefined,
      collectionId = "",
    )

  fun getProperty(key: String): String = properties[key]?.toString() ?: ""
}
